workspace/template.ts

```diff
@@ -1,7 +1,73 @@
+
+const readline = require('readline');
+const request = require('request');
+
+import * as _ from "lodash"
+import {Bus} from './bus'
+import {Coordinates} from './coordinates'
+import {PostCode} from './postCode'
+import {BusStop} from './busStop'
+import {Result} from './result'
+
 export class Template {
+    
+public handleUrlInput(urlPostcodeString: string):Promise<Result[]>{
+    var postCodeObject = new PostCode(urlPostcodeString);
+    let promiseCoordinates = postCodeObject.getCoordinates();
+    return promiseCoordinates.then((coords) => {
+        let promiseBusStops = coords.getStopPoints();
+        return promiseBusStops.then((busStopsArray) => {
+            var promisesArray: Promise<Bus[]>[] = [];
+            for(var i = 0; i < busStopsArray.length; ++i){
+                let promiseBuses = busStopsArray[i].getBuses();
+                promisesArray.push(promiseBuses);
+            }
+            return Promise.all(promisesArray).then(values => {
+                var resultsArray: Result[] = [];
+                for(var j = 0; j < values.length; ++j) {
+                    resultsArray.push(new Result(busStopsArray[j].name, values[j]))
+                }
+                return resultsArray;
+            });
+        });
+    });
+}
+    
+    private handleUserInput(){
+        var rl = readline.createInterface({
+            input: process.stdin,
+            output: process.stdout
+        });
+        let promiseInputListCommand: Promise<string> = new Promise((resolve) => { 
+            rl.question("Input a postcode\n", function(data) {
+                resolve(data.toString());
+                rl.close();
+            });
+        });
+        promiseInputListCommand.then((inputString: string) => {
+            var postCodeObject = new PostCode(inputString);
+            let promiseCoordinates = postCodeObject.getCoordinates();
+            promiseCoordinates.then((coords) => {
+                let promiseBusStops = coords.getStopPoints();
+                promiseBusStops.then((busStopsArray) => {
+                    for(var i = 0; i < busStopsArray.length; ++i){
+                        let promiseBuses = busStopsArray[i].getBuses();
+                        promiseBuses.then((busesArray) => {
+                            for(var j = 0; j < busesArray.length; ++j){
+                                var b: Bus = busesArray[j];
+                                console.log(b.destination + " " + b.line + " " + b.timeToArrival);
+                            }
+                        });
+                    }
+                });
+            });
```

Comment:
It might be nice to reduce the nesting a bit (where possible) by chaining promises instead, but I'm not going to make you redo it. Basically, the more nested things get, the more complicated they get (as I'm sure you've found!) so it you can get just a list of promises (& thens) on the same level.